File: my_code/DSA/PYTHON/DS/linked lists/singly-linked-list/singly-linked-list [Just Head].py
class Node:
    def __init__(self, val):
        self.val = val
        self.next = None

# beginning => head
# no tail and no length
class SinglyLinkedList:
    def __init__(self):
        self.head = None

    # pushing to the end of the list O(n)
    def push(self, val):
        #  Receive a value and make a new node
        new_node = Node(val)

        # if there's no head => empty list
        # so set the head to be the new Node
        if not self.head:
            self.head = new_node
            return self

        # if not empty, loop until the end
        current = self.head
        while current.next:
            current = current.next

        # then make the last node point into the new node
        current.next = new_node
        return self
    

    # remove from the end of the list O(n)
    def pop(self):

        if not self.head:
            return None

        # if only one node
        if not self.head.next:
            val = self.head.val
            self.head = None
            return val

        # use two pointers to reach the last and second last nodes
        current = self.head
        while current.next and current.next.next:
            current = current.next

        # store the last node's value
        val = current.next.val  
        # and make the node before last point to None
        # remove last node
        current.next = None 
        return val
    
    
    # remove the first element (head) O(1)
    def shift(self):
        if not self.head:
            return None

        # store the head to return it later
        val = self.head.val
        # then make the head to be the second node
        self.head = self.head.next
        return val

    # add to the beginning (head) O(1)
    def unshift(self, val):
        # make the node
        new_node = Node(val)
        # make the new node to point into the original head
        new_node.next = self.head
        # and set the head to be the new node
        self.head = new_node
        return self
    
    # get node at a specific index O(n)
    def get(self, index):
        current = self.head
        # make counter to loop until reach the counter - 1
        counter = 0

        # if there's head so we still in the list so keep going
        while current:
            # if we reach the index we want so return it
            if counter == index:
                return current
            current = current.next
            counter += 1

        # if we didn't return any value so, index out of bounds
        return None
    
    # update the value of a node at given index
    def set(self, index, val):
        # use the get method 🩷
        found_node = self.get(index)
        #  change it's value, if exist not None
        if found_node:
            found_node.val = val
            return True
        return False
    
    # insert at specific index O(n)
    def insert(self, index, val):
        #  if the index is zero so add at first
        # unshift
        if index == 0:
            return self.unshift(val)

        # else implement our insert method
        # we will use the get method, but don't need the index itself
        # I want the previous one
        prev = self.get(index - 1)
        # to ensure that the index is valid
        if not prev:
            return False
        # make a new node
        new_node = Node(val)
        # make it point the next of the previous
        new_node.next = prev.next
        # and set the previous to point on the new node
        prev.next = new_node
        return True
    
        # remove at specific index O(n)
    
    def remove(self, index):
        # if the first one shift it
        if index == 0:
            return self.shift()

        prev = self.get(index - 1)
        # ensure that the index is valid
        if not prev or not prev.next:
            return None
        
        # and set the prev next
        # to be the next of the next node
        removed = prev.next
        prev.next = removed.next

        return removed.val

    # reverse the list O(n)
    def reverse(self):
        # previous as null to make the first node point to the None
        # make it the last node
        prev = None
        #  set the current (as temp variable) to be the head    
        current = self.head

        while current:
            # in each loop store the next node to not lost the rest of the list
            next_node = current.next
            # then set the current to point to the previous node
            current.next = prev
            # then move the previous node to be the current
            prev = current
            # and move the current to the next node, which we stored it before
            # to process the next one
            current = next_node

        # finally set new head
        self.head = prev  
        return self
    
    def __str__(self):
        res = ''
        current = self.head
        while current:
            res += str(current.val) + ' '
            current = current.next
        return res
    
linked_list = SinglyLinkedList()
linked_list.push(10).push(20).push(30).push(40).push(50)
print(linked_list)
print('=' * 5)
print(linked_list.reverse())

File: my_code/DSA/JS/DS/linked-lists/singly-linked-list/singly-linked-list [Just Head].js
class Node {
  constructor(val) {
    this.val = val;
    this.next = null; // always start with null
  }
}

// beginning => head
// no tail and no length
class SinglyLinkedList {
  constructor() {
    this.head = null;
  }

  // pushing to the end of the list O(n)
  push(val) {
    // Receive a value and make a new node
    const newNode = new Node(val);

    // if there's no head => empty list
    // so set the head to be the new Node
    if (!this.head) {
      this.head = newNode;
      return this;
    }

    // if not empty, loop until the end
    let current = this.head;
    while (current.next) {
      current = current.next;
    }

    // then make the last node point into the new node
    current.next = newNode;
    return this;
  }

  // remove from the end of the list O(n)
  pop() {
    if (!this.head) return null;

    // if only one node
    if (!this.head.next) {
      const val = this.head.val;
      this.head = null;
      return val;
    }

    // use two pointers to reach the last and second last nodes
    let current = this.head;
    while (current.next && current.next.next) {
      current = current.next;
    }

    // store the last node's value
    const val = current.next.val;
    // and make the node before last point to null
    current.next = null;
    return val;
  }

  // remove the first element (head) O(1)
  shift() {
    if (!this.head) return null;

    // store the head to return it later
    const val = this.head.val;
    // then make the head to be the second node
    this.head = this.head.next;
    return val;
  }

  // add to the beginning (head) O(1)
  unshift(val) {
    // make the node
    const newNode = new Node(val);
    // make the new node point to the original head
    newNode.next = this.head;
    // and set the head to be the new node
    this.head = newNode;
    return this;
  }

  // get node at a specific index O(n)
  get(index) {
    let current = this.head;
    let counter = 0;

    // if there's head so we still in the list so keep going
    while (current) {
      // if we reach the index we want, return it
      if (counter === index) return current;
      current = current.next;
      counter++;
    }

    // if we didn't return any value, index is out of bounds
    return null;
  }

  // update the value of a node at given index
  set(index, val) {
    // use the get method 🩷
    const foundNode = this.get(index);
    // change its value, if it exists (not null)
    if (foundNode) {
      foundNode.val = val;
      return true;
    }
    return false;
  }

  // insert at specific index O(n)
  insert(index, val) {
    // if the index is zero, so add at first (unshift)
    if (index === 0) {
      this.unshift(val);
      return true;
    }

    // use the get method to get the previous node
    const prev = this.get(index - 1);
    // to ensure the index is valid
    if (!prev) return false;

    // make a new node
    const newNode = new Node(val);
    // make it point to the next of the previous
    newNode.next = prev.next;
    // and set the previous to point to the new node
    prev.next = newNode;
    return true;
  }

  // remove at specific index O(n)
  remove(index) {
    // if the first one, shift it
    if (index === 0) return this.shift();

    const prev = this.get(index - 1);
    // ensure the index is valid
    if (!prev || !prev.next) return null;

    // remove the node
    const removed = prev.next;
    prev.next = removed.next;
    return removed.val;
  }

  // reverse the list O(n)
  reverse() {
    // previous as null to make the first node point to null
    // make it the last node
    let prev = null;
    // set the current (as temp variable) to be the head
    let current = this.head;

    while (current) {
      // store the next node to not lose the rest of the list
      const nextNode = current.next;
      // set the current to point to the previous node
      current.next = prev;
      // move the previous to the current
      prev = current;
      // move the current to the next node
      current = nextNode;
    }

    // finally set new head
    this.head = prev;
    return this;
  }

  // method to return the length of the list
  length() {
    let counter = 0;
    let current = this.head;

    while (current) {
      counter += 1;
      current = current.next;
    }
    return counter;
  }

  // print
  print() {
    let res = "";
    let current = this.head;
    while (current) {
      res += current.val + " ";
      current = current.next;
    }
    console.log(res);
  }
}

// ======= Testing the list =======
let list = new SinglyLinkedList();
list.push(10).push(20).push(30).push(40).push(50);
console.log(list.remove(1));
console.log(list.length());
list.print();
